### PR TITLE
Treemap label size and direction

### PR DESCRIPTION
**Pull request by kbeame**
Using Evan's code form Dex the font size and the font direction change depending on the size of the leaf or tile's width in comparison to the string width.

### CODE REVIEW

File: src/components/treemap.tsx [hunk 1]
```diff
@@ -1,11 +1,10 @@
-import {
-  treemap,
-  treemapResquarify,
-} from 'd3';
+import { treemap, treemapResquarify } from 'd3';
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: How come one line? Did the linter say something?
I usually start doing a vertical list if there is more than one, except for certain cases where I know that there will only be 2 items brought in from a module.

File: src/components/treemap.tsx [hunk 2]
```diff
@@ -1,11 +1,10 @@
...
+import { max } from "d3-array";
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: I like that we're starting to use the more specific submodules! Nice!
This does look like Intellij may have brought it in though, in which case, we'll want to 'single-quote' the `'d3-array'` part.

File: src/components/treemap.tsx [hunk 3]
```diff
@@ -19,19 +18,20 @@ interface LayoutOptions {
 
 interface TreemapProps {
   colorScale: (input: number | string) => string;
-  data: any,
+  data: any;
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: Nice TS syntax catches!

File: src/components/treemap.tsx [hunk 4]
```diff
@@ -80,19 +80,48 @@ export default class Treemap extends React.Component<
...
     />
-  ));
+  ))
+
+    sizingProperties = d => {
+        const [ minValue, maxValue ] = this.props.fontSize;
+        const x = max([0, Math.floor(((d.x1 - d.x0) - 8))]);
+        const y = max([0, Math.floor(((d.y1 - d.y0) - 8))]);
+        const label = d.data.location_name;
+        const width = stringWidth(label, 12) / 12;
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: What's `12`? And what's `8` above here? I think it'd be cool to have these listed as CONSTANTS at the top of the file, or in a separate imported module if it get too cluttered. Thoughts?

File: src/components/treemap.tsx [hunk 5]
```diff
@@ -80,19 +80,48 @@ export default class Treemap extends React.Component<
...
+    fontDirection = d => {
+        const { direction } = this.sizingProperties(d);
+        const orientation = {
+            leftRight: `translate(3px, ${this.fontSize(d)}px) rotate(0)`,
+            topBottom: `translate(${this.fontSize(d) / 3}px, 3px) rotate(90deg)`,
+        };
+        return (direction === 'leftRight') ? orientation.leftRight : orientation.topBottom;
+    }
 
   renderText = (d) => ((
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: I don't think it's something we need to do in this PR, but it'd probably good to break this `<text />` element into it's own wrapper module soon. Maybe `<DynamicText />` or something...dealer's choice on the name.

It'll make animating the component a bit more bite size as well (I think) when we get to that step.

File: src/components/treemap.tsx [hunk 6]
```diff
@@ -80,19 +80,48 @@ export default class Treemap extends React.Component<
...
+        const x = max([0, Math.floor(((d.x1 - d.x0) - 8))]);
+        const y = max([0, Math.floor(((d.y1 - d.y0) - 8))]);
+        const label = d.data.location_name;
+        const width = stringWidth(label, 12) / 12;
+        const direction = (y > x && (width * (minValue + 1)) > x) ? 'topBottom' : 'leftRight';
+        return { x, y, width, direction, minValue, maxValue };
+    }
+
+    fontSize = d => {
+        const { x, y, width, direction, minValue, maxValue } = this.sizingProperties(d);
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: Unless you have an aversion to it, I think listing these vertically would be easier to add to. It also follows the team's current convention more. I know the airbnb lint rules yell at me when I list more than one.

File: src/components/treemap.tsx [hunk 7]
```diff
@@ -80,19 +80,48 @@ export default class Treemap extends React.Component<
...
       style={{
-        fontSize: FONT_SIZE,
-        transform: `translate(3px, ${FONT_SIZE}px)`,
+        fontSize: this.fontSize(d),
+        transform: this.fontDirection(d),
       }}
     >
       {d.data[this.props.fieldAccessors.label]}
     </text>
-  ));
```
Comment (jimvermillion) [inline src/components/treemap.tsx]: Smort, thanks for fixing my unneeded semi-colons!

File: src/utils/index.tsx [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import { stringWidth } from "./string-width";
```
Comment (jimvermillion) [inline src/utils/index.tsx]: single-quote here, probz IntelliJ again

File: src/utils/string-width.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+export const stringWidth = (stringThing, size = 12) => {
```
Comment (jimvermillion) [inline src/utils/string-width.ts]: If we do port this to IHME-UI, I think this function is already there. Aaaaand this `if` clause could be a big improvement to what is there! Nice find!

File: src/components/treemap.tsx (hunk 4 above)
Comment (kbeame) [reply src/components/treemap.tsx]: Agreed- this was stuff from Evan's code that I left in so I'm totally open to making it prettier

**Comment by jimvermillion** [comment]
Awesome! Thanks, Kat!

**Pull request merged**